cdap-app-fabric/src/main/java/io/cdap/cdap/app/store/ScanApplicationsRequest.java

```diff
@@ -0,0 +1,164 @@
+  private final ApplicationId restartAfter;
+  private final List<ApplicationFilter> filters;
+  private final SortOrder sortOrder;
+
+  private ScanApplicationsRequest(@Nullable NamespaceId namespaceId,
+                                  @Nullable ApplicationId restartAfter,
+                                  List<ApplicationFilter> filters,
+                                  SortOrder sortOrder) {
+    this.namespaceId = namespaceId;
+    this.restartAfter = restartAfter;
+    this.filters = filters;
+    this.sortOrder = sortOrder;
+  }
+
+  /**
+   *
+   * @return namespace to return applications for or null for all namespaces
+   */
+  @Nullable
+  public NamespaceId getNamespaceId() {
+    return namespaceId;
+  }
+
+  /**
+   *
+   * @return application id to start scan from (exclusive)
+   */
+  @Nullable
+  public ApplicationId getRestartAfter() {
+    return restartAfter;
+  }
+
+  /**
+   *
+   * @return additional filters to apply. All filters must be satisfied (and operation). For performance reasons
+   * it's better to put {@link ApplicationFilter.ArtifactIdFilter} first.
+   */
+  public List<ApplicationFilter> getFilters() {
+    return filters;
+  }
+
+  /**
+   *
+   * @return sort order of the results. Results are sorted by namespace, then application id in the Ascending
+   * or Descending order.
+   */
+  public SortOrder getSortOrder() {
+    return sortOrder;
+  }
+
+  @Override
+  public String toString() {
+    return "ScanApplicationsRequest{" +
+      "namespaceId=" + namespaceId +
+      "restartAfter=" + restartAfter +
+      "filters=" + filters +
+      "sortOrder=" + sortOrder +
+      '}';
+  }
+
+  /**
+   *
+   * @return builder to create a new {@link ScanApplicationsRequest}
+   */
+  public static Builder builder() {
+    return new Builder();
+  }
+
+  public static class Builder {
+    @Nullable
+    private NamespaceId namespaceId;
+    @Nullable
+    private ApplicationId restartAfter;
+    private List<ApplicationFilter> filters = new ArrayList<>();
+    private SortOrder sortOrder = SortOrder.ASC;
+
+    /**
+     * @param namespaceId namespace to scan in
+     */
+    public Builder setNamespaceId(NamespaceId namespaceId) {
+      this.namespaceId = namespaceId;
+      return this;
+    }
+
+    /**
+     *
+     * @param restartAfter restart the scan after specific application id. Useful for pagination. If namespace id
+     *                     is set, application id must be within same namespace.
+     */
+    public Builder setRestartAfter(ApplicationId restartAfter) {
+      this.restartAfter = restartAfter;
+      return this;
+    }
+
+    /**
+     *
+     * @param filter adds a filter
+     */
+    public Builder addFilter(ApplicationFilter filter) {
+      this.filters.add(filter);
+      return this;
+    }
+
+    /**
+     *
+     * @param filters adds multiple filters
+     */
+    public Builder addFilters(Collection<ApplicationFilter> filters) {
+      this.filters.addAll(filters);
```

Comment:
do we want to handle the case when `filters` is null ?

Reply:
No, null collection is a coding error and should produce a NPE